pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/operator/AggregateOperatorTest.java

```diff
@@ -18,26 +18,208 @@
 import org.testng.Assert;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeMethod;
 import org.testng.annotations.Test;
 
+import static org.apache.pinot.common.utils.DataSchema.ColumnDataType.DOUBLE;
+import static org.apache.pinot.common.utils.DataSchema.ColumnDataType.INT;
 
 
 public class AggregateOperatorTest {
 
+  private AutoCloseable _mocks;
+
+  @Mock
+  private Operator<TransferableBlock> _input;
+
+  @BeforeMethod
+  public void setUp() {
+    _mocks = MockitoAnnotations.openMocks(this);
+  }
+
+  @AfterMethod
+  public void tearDown()
+      throws Exception {
+    _mocks.close();
+  }
+
+  @Test
+  public void shouldHandleUpstreamErrorBlocks() {
+    // Given:
+    List<RexExpression> calls = ImmutableList.of(getSum(new RexExpression.InputRef(1)));
+    List<RexExpression> group = ImmutableList.of(new RexExpression.InputRef(0));
+
+    Mockito.when(_input.nextBlock())
+        .thenReturn(TransferableBlockUtils.getErrorTransferableBlock(new Exception("foo!")));
+
+    DataSchema outSchema = new DataSchema(new String[]{"sum"}, new ColumnDataType[]{DOUBLE});
+    AggregateOperator operator = new AggregateOperator(_input, outSchema, calls, group);
+
+    // When:
+    TransferableBlock block1 = operator.nextBlock(); // build
+
+    // Then:
+    Mockito.verify(_input, Mockito.times(1)).nextBlock();
+    Assert.assertTrue(block1.isErrorBlock(), "Input errors should propagate immediately");
+  }
+
+  @Test
+  public void shouldHandleEndOfStreamBlockWithNoOtherInputs() {
+    // Given:
+    List<RexExpression> calls = ImmutableList.of(getSum(new RexExpression.InputRef(1)));
+    List<RexExpression> group = ImmutableList.of(new RexExpression.InputRef(0));
+
+    Mockito.when(_input.nextBlock())
+        .thenReturn(TransferableBlockUtils.getEndOfStreamTransferableBlock());
+
+    DataSchema outSchema = new DataSchema(new String[]{"sum"}, new ColumnDataType[]{DOUBLE});
+    AggregateOperator operator = new AggregateOperator(_input, outSchema, calls, group);
+
+    // When:
+    TransferableBlock block = operator.nextBlock();
+
+    // Then:
+    Mockito.verify(_input, Mockito.times(1)).nextBlock();
+    Assert.assertTrue(block.isEndOfStreamBlock(), "EOS blocks should propagate");
+  }
+
+  @Test
+  public void shouldHandleUpstreamNoOpBlocksWhileConstructing() {
+    // Given:
+    List<RexExpression> calls = ImmutableList.of(getSum(new RexExpression.InputRef(1)));
+    List<RexExpression> group = ImmutableList.of(new RexExpression.InputRef(0));
+
+    DataSchema inSchema = new DataSchema(new String[]{"group", "arg"}, new ColumnDataType[]{INT, INT});
+    Mockito.when(_input.nextBlock())
+        .thenReturn(block(inSchema, new Object[]{1, 1}))
+        .thenReturn(TransferableBlockUtils.getNoOpTransferableBlock());
+
+    DataSchema outSchema = new DataSchema(new String[]{"sum"}, new ColumnDataType[]{DOUBLE});
+    AggregateOperator operator = new AggregateOperator(_input, outSchema, calls, group);
```

Comment:
nit: these can be private static? 
good thing is then you can use these to mock the operators. 
but i am also ok if the goal is to keep these tests isolated and easy to maintain.

Reply:
> but i am also ok if the goal is to keep these tests isolated and easy to maintain.

that's my goal :)